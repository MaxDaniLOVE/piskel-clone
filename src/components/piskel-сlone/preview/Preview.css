.preview-canvas-wrapper{
  display: flex;
  flex-direction: column;
}
.preview-canvas{
  height: 200px;
  width: 200px;
  box-shadow: 0px 0px 25px 0px rgba(0,0,0,0.75);
  image-rendering: pixelated;
  background: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAMAAABEpIrGAAAABlBMVEVMTExVVVUnhsEkAAAAHUlEQVR4AWOAAUYoQOePEAUj3v9oYDQ9gMBoegAAJFwCAbLaTIMAAAAASUVORK5CYII=) repeat;
}
.fps-change-wrapper{
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: #444;
  height: 50px;
  border-bottom-left-radius: 0.25rem;
  border-bottom-right-radius: 0.25rem;
}
.fps-value{
  display: inline-block;
  width: 30%;
  margin-bottom: 0;
  text-align: center;
}
.fps-range {
  -webkit-appearance: none;
  appearance: none;
  height: 2px;
  background: #fff;
  outline: none;
}
.fps-range::-webkit-slider-thumb {
  -webkit-appearance: none;
  appearance: none;
  width: 8px;
  height: 25px;
  border-radius: 0.25rem;
  background: #00bc8c;
  cursor: pointer;
}