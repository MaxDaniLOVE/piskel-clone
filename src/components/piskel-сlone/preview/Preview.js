import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './Preview.css';

export default class Preview extends Component {
  constructor() {
    super();
    this.state = {
      fps: 7,
    };
    this.preview = this.preview.bind(this);
    this.changeFps = this.changeFps.bind(this);
  }

  componentDidUpdate() {
    clearInterval(this.animationCountdown);
    this.preview();
  }

  preview() {
    const { frames } = this.props;
    const { fps } = this.state;
    const { previewCanvas } = this.refs;
    const context = previewCanvas.getContext('2d');
    const animation = (frame) => {
      const { image } = frame;
      const img = new Image();
      img.src = image;
      context.clearRect(0, 0, previewCanvas.width, previewCanvas.height);
      img.onload = () => {
        context.drawImage(img, 0, 0);
      };
    };
    let counter = 0;
    this.animationCountdown = setInterval(() => {
      animation(frames[counter % frames.length]);
      counter += 1;
    }, 1000 / fps);
  }

  changeFps(newFps) {
    this.setState(() => ({
      fps: newFps,
    }));
  }

  render() {
    const { resolution } = this.props;
    const { fps } = this.state;
    return (
      <div className="preview-canvas-wrapper">
        <canvas
          ref="previewCanvas"
          className="preview-canvas"
          width={resolution}
          height={resolution}
          onClick={() => {
            const { previewCanvas } = this.refs;
            previewCanvas.webkitRequestFullscreen();
          }}
        />
        <div className="fps-change-wrapper">
          <p className="text-success fps-value">{`${fps} FPS`}</p>
          <input
            type="range"
            className="fps-range"
            id="customRange1"
            min="1"
            max="24"
            defaultValue={fps}
            onChange={(event) => {
              this.changeFps(+event.target.value);
            }}
          />
        </div>
      </div>
    );
  }
}

Preview.propTypes = {
  frames: PropTypes.array.isRequired,
};
