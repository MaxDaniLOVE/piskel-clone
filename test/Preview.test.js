import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Preview from '../src/components/piskel-сlone/preview/Preview';

configure({adapter: new Adapter()});
describe('Preview', () => {
  const output = shallow(<Preview />);
  it('should set default 7 FPS', () => {
    expect(output.state().fps).toEqual(7);
  });
  it('should render correctly', () => {
    expect(shallowToJson(output)).toMatchSnapshot();
  });
});
