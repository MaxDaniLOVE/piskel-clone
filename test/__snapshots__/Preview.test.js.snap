// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preview should render correctly 1`] = `
<div
  className="preview-canvas-wrapper"
>
  <canvas
    className="preview-canvas"
    onClick={[Function]}
  />
  <div
    className="fps-change-wrapper"
  >
    <p
      className="text-success fps-value"
    >
      7 FPS
    </p>
    <input
      className="fps-range"
      defaultValue={7}
      id="customRange1"
      max="24"
      min="1"
      onChange={[Function]}
      type="range"
    />
  </div>
</div>
`;
